geoapps_utils/numerical.py

```diff
@@ -71,3 +72,176 @@ def traveling_salesman(locs: np.ndarray) -> np.ndarray:
+                vertices[edge[-2]][0] - vertices[connecting_point][0],
+                vertices[edge[-2]][1] - vertices[connecting_point][1],
+            ]
+            vec2 = [
+                vertices[new_point][0] - vertices[connecting_point][0],
+                vertices[new_point][1] - vertices[connecting_point][1],
+            ]
+            new_edge = np.concatenate((edge, [new_point]))
+
+        angle = np.arccos(
+            np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))
+        )
+
+        if angle < min_angle:
+            continue
+
+        possible_edges.append(new_edge)
+
+    return possible_edges
+
+
+def combine_edges(
+    current_edge: np.ndarray,
+    full_edges: np.ndarray,
+    vertices: np.ndarray,
+    ids: np.ndarray,
+    min_angle: float,
+    min_length: int,
+) -> list[np.ndarray]:
+    """
+    Combine edges into a single curve.
+
+    :param current_edge: Starting curve to add an edge to.
+    :param full_edges: All possible edges to add to current edge. Length 2 array.
+    :param vertices: Vertices of points.
+    :param ids: Ids of points.
+    :param min_angle: Minimum angle between points in a curve, in radians.
+    :param min_length: Minimum number of points in a curve.
+
+    :return: List of complete possible curves.
+    """
+    return_edges = []
+    # Find all edges connected to this edge
+    edge_bounds = {current_edge[0], current_edge[-1]}
+    inds = np.array([len(edge_bounds & set(edg)) == 1 for edg in full_edges])
+    connected_edges = full_edges[inds]
+
+    # Remove edges with duplicate ids and with invalid angles
+    possible_edges = filter_edges(
+        vertices, current_edge, connected_edges, ids, min_angle
+    )
+
+    if len(possible_edges) == 0 and len(current_edge) >= min_length:
+        return current_edge
+    for possible_edge in possible_edges:
+        new_edges = combine_edges(
+            possible_edge, full_edges, vertices, ids, min_angle, min_length
+        )
+        if len(new_edges) > 0 and isinstance(new_edges[0], np.intc):
+            return_edges.append(possible_edge)
+        else:
+            return_edges += new_edges
+
+    return return_edges
+
+
+def find_curves(  # pylint: disable=too-many-locals
+    vertices: np.ndarray,
+    ids: np.ndarray,
+    min_length: int,
+    max_distance: float,
+    min_angle: float,
+) -> list[list[list[float]]]:
+    """
+    Find curves in a set of points.
+
+    :param vertices: Vertices for points.
+    :param ids: Ids for points.
+    :param min_length: Minimum number of points in a curve.
+    :param max_distance: Maximum distance between points in a curve.
+    :param min_angle: Minimum angle between points in a curve, in radians.
+
+    :return: List of curves.
+    """
+    tri = Delaunay(vertices, qhull_options="QJ")
+    if not hasattr(tri, "simplices"):
+        return []
+
+    edges = np.vstack(
+        (
+            tri.simplices[:, :2],  # pylint: disable=no-member
```

Comment:
to avoid this pylint disable, configure pylintrc as in geoapps (search there for Delaunay)